optuna/_deprecated.py

```diff
@@ -0,0 +1,141 @@
+import functools
+import inspect
+from packaging import version
+import textwrap
+from typing import Any
+from typing import Callable
+import warnings
+
+from optuna._experimental import _get_docstring_indent
+from optuna._experimental import _validate_version
+
+
+_DEPRECATION_NOTE_TEMPLATE = """
+
+.. note::
+    Deprecated in v{d_ver}. This feature will be removed in the future. The removal of this
+    feature is currently scheduled for v{r_ver}, but this schedule is subject to change.
+    See https://github.com/optuna/optuna/releases/tag/v{d_ver}.
+"""
+
+
+def _validate_two_version(old_version: str, new_version: str) -> None:
+    if version.parse(old_version) > version.parse(new_version):
+        raise ValueError(
+            "Invalid version relationship. The deprecated version must be smaller than "
+            "the removed version, but (deprecated version, removed version) = ({}, {}) are "
+            "specified.".format(old_version, new_version)
+        )
+
+
+def _formatting_text(note: str) -> str:
+    if note == "":
+        return note
+    else:
+        return "".join(["    " + s + "\n" for s in note.split(sep="\n") if len(s) > 0])
+
+
+def deprecated(
+    deprecated_version: str, removed_version: str, name: str = None, additional_text: str = ""
```

Comment:
[nit-picking]
I guess most of type hintings come from my terrible `experimental` implementation, though, if you allow `None` explicitly for `name` argument, then it should be `Optional[str]`. Personally empty string is enough as the default for `name`.
```
    deprecated_version: str, removed_version: str, name: Optional[str] = None, additional_text: str = ""
```

or 

```
    deprecated_version: str, removed_version: str, name: str = "", additional_text: str = ""
```

Reply:
Thanks a lot! Fixed.